components/locale/src/langid.rs

```diff
@@ -0,0 +1,151 @@
+/// assert_eq!(li.script.unwrap(), "Latn");
+/// assert_eq!(li.region.unwrap(), "US");
+/// assert_eq!(li.variants.get(0).unwrap(), "valencia");
+/// ```
+///
+/// [`Unicode Language Identifier`]: https://unicode.org/reports/tr35/tr35.html#Unicode_language_identifier
+#[derive(Default, Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord)]
+pub struct LanguageIdentifier {
+    /// Language subtag of the LanguageIdentifier
+    pub language: subtags::Language,
+    /// Script subtag of the LanguageIdentifier
+    pub script: Option<subtags::Script>,
+    /// Region subtag of the LanguageIdentifier
+    pub region: Option<subtags::Region>,
+    /// Variant subtags of the LanguageIdentifier
+    pub variants: subtags::Variants,
+}
+
+impl LanguageIdentifier {
+    /// A constructor which takes a utf8 slice, parses it and
+    /// produces a well-formed `LanguageIdentifier`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_locale::LanguageIdentifier;
+    ///
+    /// let li = LanguageIdentifier::from_bytes(b"en-US")
+    ///     .expect("Parsing failed.");
+    ///
+    /// assert_eq!(li.to_string(), "en-US");
+    /// ```
+    pub fn from_bytes(v: &[u8]) -> Result<Self, ParserError> {
+        parse_language_identifier(v, ParserMode::LanguageIdentifier)
+    }
+
+    /// A constructor which takes a utf8 slice which may contain extension keys,
+    /// parses it and produces a well-formed `LanguageIdentifier`.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_locale::LanguageIdentifier;
+    ///
+    /// let li = LanguageIdentifier::from_locale_bytes(b"en-US-x-posix")
+    ///     .expect("Parsing failed.");
+    ///
+    /// assert_eq!(li.to_string(), "en-US");
+    /// ```
+    ///
+    /// This method should be used for input that may be a locale identifier.
+    /// All extensions will be lost.
+    pub fn from_locale_bytes(v: &[u8]) -> Result<Self, ParserError> {
+        parse_language_identifier(v, ParserMode::Locale)
+    }
+
+    /// This is a best-effort operation that performs all available levels of canonicalization.
+    ///
+    /// At the moment the operation will normalize casing and the separator, but in the future
+    /// it may also validate and update from deprecated subtags to canonical ones.
+    ///
+    /// # Example
+    ///
+    /// ```
+    /// use icu_locale::LanguageIdentifier;
+    ///
+    /// assert_eq!(LanguageIdentifier::canonicalize("pL_latn_pl"), Ok("pl-Latn-PL".to_string()));
+    /// ```
+    pub fn canonicalize<S: AsRef<[u8]>>(input: S) -> Result<String, ParserError> {
+        let lang_id = Self::from_bytes(input.as_ref())?;
+        Ok(lang_id.to_string())
+    }
+}
+
+impl FromStr for LanguageIdentifier {
+    type Err = ParserError;
+
+    fn from_str(source: &str) -> Result<Self, Self::Err> {
+        Self::from_bytes(source.as_bytes())
+    }
+}
+
+impl std::fmt::Display for LanguageIdentifier {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
```

Comment:
Is this only for debug output? It's pulling in std:: - do we want to do that?

Reply:
It's not for [`Debug`](https://doc.rust-lang.org/std/fmt/trait.Display.html). It's for [`Display`](https://doc.rust-lang.org/std/fmt/trait.Display.html).
You can think of `Display` as a well designed `ToString` ;)

It uses `std` for now, but it's part of [`core`](https://doc.rust-lang.org/core/fmt/trait.Display.html) as well and [`alloc`](https://doc.rust-lang.org/alloc/fmt/trait.Display.html).
Once @sffc comes to conclusion about `no_std` we will decide what to do with this, but it should be forward compatible for all scenarios in which we have `alloc`. If we don't have alloc, then well, we can't construct a `String`, and then we'll have to put this behind a `feature` cfg flag.

Reply:
So, in the `no_alloc` feature future scenario (pun intended), we'd allow people to construct this component without ability to stringify the `LanguageIdentifier`, and that's a reasonable solution.
I expect any `no_alloc` user to not be interested in stringifiaction of the identifiers.

So, adding it allows us to present the `Display` mode (which is neat because it takes a formatter, so it can write to a pre-allocated string!), which is useful for the regular consumers, and then when/if we'll be adding a `no_alloc` mode we'll put this trait behind a flag.